openedx/core/lib/block_structure/transformers.py

```diff
@@ -105,3 +108,50 @@ def is_collected_outdated(cls, block_structure):
             )
 
         return bool(outdated_transformers)
+
+    def _categorize_transformers(self):
+        """
+        Returns a tuple of
+            1. list of transformers that implement the transform_block_filter
+            method, and
+            2 list of the rest of the transformers
+        """
+        with_block_filters = []
+        other = []
+        for transformer in self._transformers:
+            if self._is_transform_block_filter_implemented(transformer):
+                with_block_filters.append(transformer)
+            else:
+                other.append(transformer)
+        return with_block_filters, other
+
+    def _is_transform_block_filter_implemented(self, transformer):
+        """
+        Returns whether the given transformer has its own implementation of
+        the transform_block_filter method.
+        """
+        return transformer.transform_block_filter.__func__ != BlockStructureTransformer.transform_block_filter.__func__
+
+    def _transform_with_block_filters(self, transformers, block_structure):
+        """
+        Transforms the given block_structure using the transform_block_filter
+        method from the given transformers.
+        """
+        if not transformers:
+            return
+        combined_filters = functools_reduce(
+            lambda t1_filter, t2_filter:
+                    t1_filter(self.usage_info, block_structure) and
+                    t2_filter(self.usage_info, block_structure),
+            [t.transform_block_filter for t in transformers],
+            lambda usage_info, block_structure: True
+        )
```

Comment:
Loving this use of `functools.reduce()`, makes this way cleaner to read than what I had before.